test-clients/src/main/java/com/hedera/services/yahcli/commands/signedstate/SignedStateHolder.java

```diff
@@ -0,0 +1,253 @@
+/*
+ * Copyright (C) 2022 Hedera Hashgraph, LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.hedera.services.yahcli.commands.signedstate;
+
+import com.hedera.services.ServicesState;
+import com.hedera.services.state.migration.AccountStorageAdapter;
+import com.hedera.services.state.virtual.VirtualBlobKey;
+import com.hedera.services.state.virtual.VirtualBlobKey.Type;
+import com.hedera.services.state.virtual.VirtualBlobValue;
+import com.swirlds.common.constructable.ConstructableRegistry;
+import com.swirlds.platform.state.signed.SignedStateFileReader;
+import com.swirlds.virtualmap.VirtualMap;
+import edu.umd.cs.findbugs.annotations.NonNull;
+import java.io.Closeable;
+import java.io.IOException;
+import java.nio.file.Path;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Collection;
+import java.util.HashSet;
+import java.util.Set;
+
+/**
+ * Navigates a signed state "file" and returns information from it
+ *
+ * <p>A "signed state" is actually a directory tree, at the top level of which is the serialized
+ * merkle tree of the hashgraph state. That is in a file named `SignedState.swh` and file is called
+ * the "signed state file". But the whole directory tree must be present.
+ *
+ * <p>This uses a `SignedStateFileReader` to suck that entire merkle tree into memory, plus the
+ * indexes of the virtual maps ("vmap"s) - ~1Gb serialized (2022-11). Then you can traverse the
+ * rematerialized hashgraph state.
+ *
+ * <p>Currently implements only operations needed for looking at contracts: - {@link
+ * #getAllKnownContracts()} looks in all the accounts to get all the contract ids present, - {@link
+ * #getAllContractContents(Collection)} returns a map, indexed by contract id, of the contract
+ * bytecodes.
+ */
+public class SignedStateHolder implements Closeable {
+
+    @NonNull private final Path swh;
+    @NonNull private ServicesState platformState;
+
+    public SignedStateHolder(@NonNull Path swhFile) throws Exception {
```

Comment:
Marking the passed in parameter as `final`, if applicable, might be useful.  Not so much in this case as it is obvious but in some of the other methods below it might be good to know if the parameter is immutable.

Reply:
Ah, spot-checking the code in `hedera-mono-service` I see now that `final` is routinely used for method parameters.  OK.  Don't see it used in _interfaces_ though.  Why's that?

Reply:
It seems to be pointless as the implementation can omit the `final` keyword.

Reply:
That's a strange Java misfeature.  The _contract_ i.e., the interface, can't commit implementations to a guarantee to the _user_ of the contract that passed reference parameters won't be modified?  Yet at the same time it refuses to allow _that_ important specification it _does_ allow the `final` keyword to be put there (in an interface, I tried it) so that the user of the contract might be misled into thinking its enforced?